beluga/include/beluga/primitives.hpp

```diff
@@ -67,27 +67,27 @@ struct has_member_state : std::false_type {};
 template <class T>
 struct has_member_state<T, std::void_t<decltype(std::declval<T>().state())>> : std::true_type {};
 
-template <typename T>
+template <class T>
 inline constexpr bool has_member_state_v = has_member_state<T>::value;
 
-/// \endcond
+template <class T, class = void>
+struct has_non_member_state : std::false_type {};
 
-/// Default state implementation.
-/**
- * Assumes that the particle is a tuple and the first element is the state.
- */
-template <typename T>
-constexpr decltype(auto) state(T&& t) noexcept(noexcept(std::get<0>(std::forward<T>(t)))) {
-  return std::get<0>(std::forward<T>(t));
-}
+template <class T>
+struct has_non_member_state<T, std::void_t<decltype(state(std::declval<T>()))>> : std::true_type {};
+
+template <class T>
+inline constexpr bool has_non_member_state_v = has_non_member_state<T>::value;
+
+/// \endcond
 
 /// Customization point object type for accessing the `state` of a particle.
 /**
  * See https://en.cppreference.com/w/cpp/ranges/cpo.
  */
 struct state_fn {
   /// Overload for when the particle type defines a member method.
-  template <typename T, std::enable_if_t<has_member_state_v<T>, int> = 0>
+  template <class T, std::enable_if_t<has_member_state_v<T>, int> = 0>
```

Comment:
@nahueespinosa meta: why support getter methods but not plain members? A `struct` with `state` and `weight` members is the simplest particle definition I can think of, even if it works against performance.

Reply:
You got it! I'll update the PR soon.

Reply:
Done in https://github.com/Ekumen-OS/beluga/pull/280/commits/5db0b4084d498269ed7d9d967212e7c2e6f88426!